File: string-calculator/strign-calculator-java/src/main/java/pl/frognar/_05/StringCalculator.java
package pl.frognar._05;

import java.util.Arrays;
import java.util.stream.Stream;

public class StringCalculator {
    public int add(String numbers) {
        if (numbers == null || numbers.isEmpty())
            return 0;
        var numberList = Arrays.stream(numbers.split(","))
                .map(Integer::parseInt);
        return calculateSumOf(numberList);
    }

    private int calculateSumOf(Stream<Integer> numbers) {
        return numbers.reduce(0, Integer::sum);
    }
}

File: string-calculator/string-calculator-py/02/string_calculator.py
from re import split
from typing import List


class NegativesNotAllowedError(RuntimeError):
    def __int__(self, message):
        super.__init__(message)


class StringCalculator:
    @classmethod
    def add(cls, numbers: str) -> int:
        if not numbers:
            return 0
        number_list = cls.split_and_convert_to_numbers(numbers)
        if any(n < 0 for n in number_list):
            raise NegativesNotAllowedError('negatives not allowed: -4')
        return cls.calculate_sum_of(number_list)

    @staticmethod
    def split_and_convert_to_numbers(numbers):
        default_delimiters = '[,\n]'
        return [int(n) for n in split(default_delimiters, numbers)]

    @staticmethod
    def calculate_sum_of(numbers: List[int]) -> int:
        return sum(numbers)
